linera-service/src/linera-exporter/storage.rs

```diff
@@ -0,0 +1,428 @@
+                chain_states_cache,
+                exporter_state_view: view,
+            },
+            exporter_storage,
+        ))
+    }
+
+    pub(super) async fn get_block(
+        &self,
+        hash: CryptoHash,
+    ) -> Result<Arc<GenericCertificate<ConfirmedBlock>>, ExporterError> {
+        self.shared_storage.get_block(hash).await
+    }
+
+    pub(super) async fn is_block_indexed(
+        &mut self,
+        block_id: &BlockId,
+    ) -> Result<bool, ExporterError> {
+        if let Some(status) = self.chain_states_cache.get(&block_id.chain_id) {
+            return Ok(status.height >= block_id.height);
+        }
+
+        if let Some(status) = self
+            .exporter_state_view
+            .get_chain_status(&block_id.chain_id)
+            .await?
+        {
+            let result = status.height >= block_id.height;
+            self.chain_states_cache.insert(block_id.chain_id, status);
+            return Ok(result);
+        }
+
+        Err(ExporterError::UnprocessedChain)
+    }
+
+    pub(super) async fn index_chain(&mut self, block_id: &BlockId) -> Result<(), ExporterError> {
+        ensure!(
+            block_id.height == BlockHeight::ZERO,
+            ExporterError::BadInitialization
+        );
+        self.shared_storage.push_block(block_id.hash).await?;
+        self.exporter_state_view
+            .initialize_chain(block_id.clone())
+            .await?;
+        self.chain_states_cache
+            .insert(block_id.chain_id, block_id.clone().into());
+
+        Ok(())
+    }
+
+    pub(super) async fn index_block(&mut self, block_id: &BlockId) -> Result<(), ExporterError> {
+        if block_id.height == BlockHeight::ZERO {
+            self.index_chain(block_id).await?;
+            return Ok(());
+        }
+
+        if self
+            .exporter_state_view
+            .index_block(block_id.clone())
+            .await?
+        {
+            self.shared_storage.push_block(block_id.hash).await?;
+            self.chain_states_cache
+                .insert(block_id.chain_id, block_id.clone().into());
+        }
+
+        Ok(())
+    }
+
+    pub(super) async fn save(&mut self) -> Result<(), ExporterError> {
+        let mut batch = Batch::new();
+        self.shared_storage
+            .shared_canonical_state
+            .flush(&mut batch)?;
+        self.exporter_state_view
+            .set_destination_states(self.shared_storage.destination_states.clone());
+        self.exporter_state_view.flush(&mut batch)?;
+        self.exporter_state_view.rollback();
+        if let Err(e) = self
+            .exporter_state_view
+            .context()
+            .store()
+            .write_batch(batch)
+            .await
+        {
+            Err(ExporterError::ViewError(e.into()))?;
+        };
+
+        Ok(())
+    }
+}
+
+struct CanonicalState<C> {
+    count: usize,
+    state_cache: Arc<FifoCache<usize, CryptoHash>>,
+    state_updates_buffer: Arc<DashMap<usize, CryptoHash>>,
```

Comment:
Can you explain why we need both `state_cache` and `state_updates_buffer`? They both work on the same keys and values, they're updated together.